src/QsCompiler/QirGeneration/FunctionContext.cs

```diff
@@ -0,0 +1,251 @@
+        /// ensuring that emitted code does not result in an invalid block.
+        /// </summary>
+        /// <remarks>
+        /// If the current block ends with a <see cref="Terminator"/>, a new (unreachable) block is
+        /// prepared, and <paramref name="func"/> is invoked with its builder. If instructions
+        /// are emitted, the unreachable block becomes the current block. Otherwise, the empty
+        /// unreachable block is removed.
+        /// <para/>
+        /// Note that nested calls to <see cref="Emit"/> are valid. In this case, validation is only
+        /// performed when the root call to <paramref name="func"/> returns.
+        /// I.e. Inner calls to <see cref="Emit"/> invoke their <paramref name="func"/> with
+        /// the same builder being used by that of the most outter.
+        /// </remarks>
+        /// <typeparam name="T">The return type of <paramref name="func"/>.</typeparam>
+        /// <param name="func">The function.</param>
+        /// <returns>The return value of <paramref name="func"/>.</returns>
+        /// <exception cref="InvalidOperationException">
+        /// The call to <paramref name="func"/> emitted an instruction sequence containing
+        /// a <see cref="Terminator"/> that is not the last instruction.
+        /// </exception>
+        internal T Emit<T>(Func<InstructionBuilder, T> func)
+        {
+            if (this.activelyEmittingTo != null)
+            {
+                return func(this.activelyEmittingTo);
+            }
+
+            // Reserve a new unique name for an unreachable block based on
+            // the name of the first preceeding reachable block.
+            string ReserveUnreachableBlockName()
+            {
+                string originalBlockName;
+                if (!this.unreachableToOriginalName.TryGetValue(this.CurrentBlock.Name, out originalBlockName))
+                {
+                    originalBlockName = this.CurrentBlock.Name;
+                }
+
+                var blockNameBase = $"{originalBlockName}__unreachable";
+                var uniqueBlockName = this.blockNameAllocate(blockNameBase);
+
+                this.unreachableToOriginalName.Add(uniqueBlockName, originalBlockName);
+                return uniqueBlockName;
+            }
+
+            try
+            {
+                T result;
+                var currentInstCount = 0;
+
+                if (this.IsCurrentBlockTerminated)
+                {
+                    // Current block has already been terminated.
+                    // Create a new block to hold any unreachable instructions emitted.
+                    // If no instructions are emitted by func(), we will delete this block.
+                    // For this reason, we do not reserve a unique block name until we know
+                    // we'll keep the block.
+                    var unreachableBlock = this.AddBlockAfterCurrent("temp__unreachable", blockNameAllocate: null);
```

Comment:
Is it possible that the temporary name may accidentally end up in the unreachable code? For example, if there is a for loop in the unreachable block, it will get generated before the name is updated. Does that end up needing to use the containing block's name?

Reply:
Heh, unit tests speak louder than words :) Thanks for adding the test cases the demonstrate the behavior here! (https://github.com/microsoft/qsharp-compiler/pull/941/commits/8d858acc0b1283862538a3d9ba695db03cfc09b2)

Reply:
I'm not sure I'm following completely, but are you referring to the case that an instruction inside an `Emit` call references the current block (e.g. `PhiNode`)? For the temporary name to be the current block's name, it would have to be read inside an `Emit`. I think even in that case, instructions like `PhiNode` take the block by reference, and not name, so I'd wager that the emitted PhiNode instruction would end up with the updated name. I think that'd be worth testing prior to merge.

Something I didn't originally consider (maybe this is actually what you were getting at) is that callers need to know that `Emit` might change the current block.

For example, in this case, `precedingBlock` should probably be defined only after the first call to `Emit`, since the block might change, which would then result in an incorrect `PhiNode`.

```c#
// End the current block by branching into the header of the loop
BasicBlock precedingBlock = this.FunctionContext.CurrentBlock;

// We need to open a scope before starting the for-loop by creating the header block, since
// it is possible for the condition to perform an allocation that needs to get cleaned up.
this.ScopeMgr.OpenScope();
this.FunctionContext.Emit(b => b.Branch(headerBlock));

// Header block: create/update phi node representing the iteration variable and evaluate the condition
this.FunctionContext.SetCurrentBlock(headerBlock);
var loopVariable = this.FunctionContext.Emit(b => b.PhiNode(this.Types.Int));
loopVariable.AddIncoming(startValue, precedingBlock);
```

This specific case doesn't cause an issue in the (newly added :)) unit test, because a bit of code gets emitted before this method is called to calculate the loop end.

I think if we do go with this approach, it's very important to have unit tests that demonstrate that all statements after a terminal one work as expected, since it'd probably be pretty easy to forget that `Emit` might change the block.

That said, if this seems too difficult to reason about, I'd completely understand. There's always the possibility of just pruning the AST of unreachable code prior to QIR generation, but I just thought it'd be cool if the generation could support this.

Reply:
> Heh, unit tests speak louder than words :)

I promise I was working on the words too! I just thought I would write some tests to see if it'd help me understand what you were saying :)